stdlib: add socks proxy to atlas client

Harshil2107 opened the pull request:
Change-Id: I683e329b4a915fa8e767ee7585eb70c5bf8d1439

Harshil2107 committed:
stdlib: add socks proxy to atlas client

Change-Id: I683e329b4a915fa8e767ee7585eb70c5bf8d1439

powerjg commented on src/python/gem5/resources/client_api/atlasclient.py:
```diff
@@ -125,6 +126,27 @@ def _atlas_http_json_req(
+
+            import socks
+
+            IP_ADDR, host_port = use_proxy.split(":")
+            PORT = int(host_port)
```
Why is `IP_ADDR` and `PORT` capitalized?

Harshil2107 replied:
It was capitalized in the original code so I did not change it.

powerjg commented on src/python/gem5/resources/client_api/atlasclient.py:
```diff
@@ -125,6 +126,27 @@ def _atlas_http_json_req(
 
+        # check to see if user requests a proxy connection
+        use_proxy = os.getenv("GEM5_USE_PROXY")
+        if use_proxy:
+            # If the "use_proxy" variable is specified we setup a socks5
+            # connection.
```
Does it make sense to split this out into a function? Is this duplicated code from the other place you check proxies?

Harshil2107 replied:
Yes, I was thinking of moving this to a function that just returns the context as we use this context twice first in atlasclient and then in downloader.py.

Harshil2107 committed:
stdlib, resources: move code for setting proxy

- Moved code for setting up SOCKS proxy in a separate function which is called only once
- Update clients to be compatible with this change

Change-Id: Ic7bf0d34ba7dc0e653bafb36a4bdafb4bf146b60

powerjg commented on src/python/gem5/resources/client_api/atlasclient.py:
```diff
@@ -133,7 +136,10 @@ def _atlas_http_json_req(
         for attempt in itertools.count(start=1):
             try:
-                response = request.urlopen(req)
+                if proxy_context is not None:
+                    response = request.urlopen(req, context=proxy_context)
+                else:
+                    response = request.urlopen(req)
```
This if statement is unneeded. You can pass `None` as the context.

```suggestion
                response = request.urlopen(req, context=proxy_context)
```

powerjg commented on src/python/gem5/resources/downloader.py:
```diff
@@ -61,7 +61,12 @@
+    url: str,
+    download_to: str,
+    max_attempts: int = 6,
+    proxy_context: object = None,
```
Is that type right?

I think it's `Optional[ssl.SSLContext]`

powerjg commented on src/python/gem5/resources/downloader.py:
```diff
@@ -181,6 +172,27 @@ def _download(url: str, download_to: str, max_attempts: int = 6) -> None:
+
+        import socks
+
+        IP_ADDR, host_port = use_proxy.split(":")
+        PORT = int(host_port)
```
Why are these all caps?

Harshil2107 replied:
They were caps in the original code, I will change them to lower case.

powerjg commented on src/python/gem5/resources/downloader.py:
```diff
@@ -215,6 +227,7 @@ def get_resource(
     clients: Optional[List] = None,
     gem5_version: Optional[str] = core.gem5Version,
     quiet: bool = False,
+    proxy_context: Optional[object] = None,
```
I'm pretty sure the type is https://docs.python.org/3/library/ssl.html#ssl.SSLContext

powerjg commented on src/python/gem5/resources/client_api/jsonclient.py:
```diff
@@ -78,6 +78,7 @@ def get_resources(
         resource_id: Optional[str] = None,
         resource_version: Optional[str] = None,
         gem5_version: Optional[str] = None,
+        proxy_context: Optional[object] = None,
```
Type is not object

powerjg commented on src/python/gem5/resources/client_api/client_wrapper.py:
```diff
@@ -104,6 +104,7 @@ def get_all_resources_by_id(
         self,
         resource_id: str,
         clients: Optional[List[str]] = None,
+        proxy_context: Optional[object] = None,
```
Type is not object

powerjg commented on src/python/gem5/resources/client_api/atlasclient.py:
```diff
@@ -158,6 +164,7 @@ def get_resources(
         resource_id: Optional[str] = None,
         resource_version: Optional[str] = None,
         gem5_version: Optional[str] = None,
+        proxy_context: Optional[object] = None,
```
Type is not object

powerjg commented on src/python/gem5/resources/client_api/atlasclient.py:
```diff
@@ -102,6 +103,7 @@ def _atlas_http_json_req(
         data_json: Dict[str, Any],
         headers: Dict[str, str],
         purpose_of_request: Optional[str],
+        proxy_context: Optional[object] = None,
```
Type should be https://docs.python.org/3/library/ssl.html#ssl.SSLContext, I think

powerjg commented on src/python/gem5/resources/client_api/abstract_client.py:
```diff
@@ -111,10 +114,15 @@ def filter_incompatible_resources(
 
-    def get_resources_by_id(self, resource_id: str) -> List[Dict[str, Any]]:
+    def get_resources_by_id(
+        self, resource_id: str, proxy_context: object
```
I don't think the type is correct

powerjg commented on src/python/gem5/resources/client_api/abstract_client.py:
```diff
@@ -58,6 +58,7 @@ def get_resources(
         resource_id: Optional[str] = None,
         resource_version: Optional[str] = None,
         gem5_version: Optional[str] = None,
+        proxy_context: Optional[object] = None,
```
Type is not correct

powerjg commented on src/python/gem5/resources/client.py:
```diff
@@ -151,6 +151,7 @@ def get_resource_json_obj(
     resource_version: Optional[str] = None,
     clients: Optional[List[str]] = None,
     gem5_version: Optional[str] = core.gem5Version,
+    proxy_context: Optional[object] = None,
```
I don't think the type is correct. I believe it should be https://docs.python.org/3/library/ssl.html#ssl.SSLContext

BobbyRBruce commented on src/python/gem5/resources/resource.py:
```diff
@@ -963,13 +966,14 @@ def obtain_resource(
     :param quiet: If ``True``, suppress output. ``False`` by default.
     """
-
+    proxy_context = get_proxy_context()
```
Why do we need to get the proxy_context here? Can't we get this in the `get_resource_json` function? Can't we just call this `get_proxy_context` function when it's needed? This approach requires changing quite a lot of function signatures and passing data around.

Harshil2107 replied:
I did this because we only want to change the context once and then pas the same context around.

BobbyRBruce replied:
If you only want to ever run this function once to get the data then I think there's an easier way. What about the following?

```py
# In downloader.py

_gem5_ssl_context = None
def get_proxy_context() -> TYPE_HERE:
    global _gem5_ssl_context
    use_proxy = os.getenv("GEM5_USE_PROXY")
    if use_proxy and not _gem5_ssl_context:
        import socket

        import socks

        ip_addr, host_port = use_proxy.split(":")
        port = int(host_port)
        socks.set_default_proxy(socks.SOCKS5, ip_addr, port)
        socket.socket = socks.socksocket

        # base SSL context for https connection
        _gem5_ssl_context = ssl.create_default_context()
        ctx.check_hostname = False
        ctx.verify_mode = ssl.CERT_NONE
    return gem5_ssl_context
```

With this we'd only need to call `get_proxy_context` at the point of creating the connection. The setting of the default proxy has once and we store the ssl_context to be returned everytime.

Harshil2107 replied:
This causes an import error due to circular imports, as the downloader.py needs to access a function from client.py and client.py imports the AltasClient which in turn import get_proxy_context() from downloader.py.

BobbyRBruce commented on src/python/gem5/resources/downloader.py:
```diff
@@ -181,6 +173,26 @@ def _download(url: str, download_to: str, max_attempts: int = 6) -> None:
             )
 
 
+def get_proxy_context():
```
Please put a return type here

Harshil2107 committed:
stdlib: address requested changes

Change-Id: Ib8821ac0881be3b60c8b36ee35174be4623fe328

Harshil2107 commented:
I implemented the suggestion you mentioned, and it causes circular imports. I think we can maybe move this context function to abstract_client.py and that should solve this issue.

Harshil2107 committed:
stdlib: moved ssl context code to utils

Change-Id: Ia1234772a4f5892ea5fbb7b6994fcabc972326cc

Harshil2107 commented:
This should fix the circular import error.

BobbyRBruce commented on src/python/gem5/utils/socks_ssl_context.py:
You need to add copyright info.

BobbyRBruce commented on src/python/gem5/utils/socks_ssl_context.py:
```diff
@@ -0,0 +1,25 @@
+
+
+def get_proxy_context() -> Optional[ssl.SSLContext]:
+    global _gem5_ssl_context
```
Can you add some docstring here?

BobbyRBruce commented on src/python/gem5/resources/downloader.py:
```diff
@@ -181,6 +170,29 @@ def _download(url: str, download_to: str, max_attempts: int = 6) -> None:
             )
 
 
+_gem5_ssl_context = None
+
+
+def get_proxy_context() -> Optional[ssl.SSLContext]:
+    global _gem5_ssl_context
+    use_proxy = os.getenv("GEM5_USE_PROXY")
+    if use_proxy and not _gem5_ssl_context:
+        import socket
+
+        import socks
+
+        ip_addr, host_port = use_proxy.split(":")
+        port = int(host_port)
+        socks.set_default_proxy(socks.SOCKS5, ip_addr, port)
+        socket.socket = socks.socksocket
+
+        # base SSL context for https connection
+        _gem5_ssl_context = ssl.create_default_context()
+        _gem5_ssl_context.check_hostname = False
+        _gem5_ssl_context.verify_mode = ssl.CERT_NONE
+    return _gem5_ssl_context
+
+
```
This is no longer needed.

BobbyRBruce commented on src/python/gem5/resources/downloader.py:
```diff
@@ -79,38 +85,21 @@ def _download(url: str, download_to: str, max_attempts: int = 6) -> None:
                 request = urllib.request.Request(url)
-                with urllib.request.urlopen(request, context=ctx) as fr:
+                with urllib.request.urlopen(
+                    request, context=proxy_context
```
```suggestion
                    request, context=get_proxy_context()
```

BobbyRBruce commented on src/python/gem5/resources/client_api/jsonclient.py:
```diff
@@ -25,6 +25,7 @@
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 import json
+import ssl
```
I wouldn't think this import is needed.

BobbyRBruce commented on src/python/gem5/resources/client_api/client_wrapper.py:
```diff
@@ -25,6 +25,7 @@
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 import itertools
+import ssl
```
I wouldn't think this import is needed.

BobbyRBruce commented on src/python/gem5/resources/client_api/atlasclient.py:
```diff
@@ -158,6 +161,7 @@ def get_resources(
         resource_id: Optional[str] = None,
         resource_version: Optional[str] = None,
         gem5_version: Optional[str] = None,
+        proxy_context: Optional[ssl.SSLContext] = None,
```
Remove this. No longer required.

BobbyRBruce commented on src/python/gem5/resources/client_api/atlasclient.py:
```diff
@@ -133,7 +136,7 @@ def _atlas_http_json_req(
         for attempt in itertools.count(start=1):
             try:
-                response = request.urlopen(req)
+                response = request.urlopen(req, context=proxy_context)
```
```suggestion
                response = request.urlopen(req, context=get_proxy_context())
```

BobbyRBruce commented on src/python/gem5/resources/client_api/abstract_client.py:
Nothing in this file should be changed.

BobbyRBruce replied:
@Harshil2107 : Please don't "Resolve conversation" when you didn't resolve the issue. There's still changes in this file.

BobbyRBruce commented on src/python/gem5/resources/client.py:
Nothing in this file should be changed.

BobbyRBruce commented on src/python/SConscript:
```diff
@@ -312,6 +314,7 @@ PySource('gem5.utils.multiprocessing',
 PySource('gem5.utils.multiprocessing',
     'gem5/utils/multiprocessing/popen_spawn_gem5.py')
 
+
```
Accidental new line here. Remove.

BobbyRBruce commented on src/python/gem5/resources/client_api/atlasclient.py:
```diff
@@ -26,6 +26,8 @@
 
 import itertools
 import json
+import os
+import ssl
```
These imports are no longer needed.

BobbyRBruce commented on src/python/gem5/resources/client_api/atlasclient.py:
```diff
@@ -123,8 +126,8 @@ def _atlas_http_json_req(
 
         **Warning**: This function assumes a JSON response.
         """
+        proxy_context = get_proxy_context()
```
Remove this.

BobbyRBruce commented on src/python/gem5/resources/client_api/atlasclient.py:
```diff
@@ -123,8 +126,8 @@ def _atlas_http_json_req(
         """
+        proxy_context = get_proxy_context()
         data = json.dumps(data_json).encode("utf-8")
-
```
Re-add the newline here.

BobbyRBruce commented on src/python/gem5/resources/downloader.py:
```diff
@@ -28,6 +28,7 @@
 import os
 import random
 import shutil
+import ssl
```
I wouldn't think this import is needed.

BobbyRBruce commented on src/python/gem5/resources/downloader.py:
```diff
@@ -61,7 +63,11 @@
 """
 
 
-def _download(url: str, download_to: str, max_attempts: int = 6) -> None:
+def _download(
+    url: str,
+    download_to: str,
+    max_attempts: int = 6,
+) -> None:
```
I don't think this needs reformatted.

BobbyRBruce commented on src/python/gem5/resources/downloader.py:
```diff
@@ -79,38 +85,21 @@ def _download(url: str, download_to: str, max_attempts: int = 6) -> None:
 
     # TODO: This whole setup will only work for single files we can get via
     # wget. We also need to support git clones going forward.
-
+    proxy_context = get_proxy_context()
```
Remove this.

Harshil2107 committed:
stdlib: Cleaned up files

Change-Id: Ibefc9b73c1672ab271a285f5630134ee14d14873

BobbyRBruce commented on src/python/gem5/resources/downloader.py:
```diff
@@ -79,38 +80,20 @@ def _download(url: str, download_to: str, max_attempts: int = 6) -> None:
 
     # TODO: This whole setup will only work for single files we can get via
     # wget. We also need to support git clones going forward.
-
```
Re-add this line to keep git clean.

BobbyRBruce commented on src/python/gem5/resources/downloader.py:
```diff
@@ -79,38 +80,20 @@ def _download(url: str, download_to: str, max_attempts: int = 6) -> None:
-                ctx = ssl.create_default_context()
-                ctx.check_hostname = False
-                ctx.verify_mode = ssl.CERT_NONE
-
                 # get the file as a bytes blob
                 request = urllib.request.Request(url)
-                with urllib.request.urlopen(request, context=ctx) as fr:
+                with urllib.request.urlopen(
+                    request, context=get_proxy_context()
+                ) as fr:
```
You could be a little neater here. Checking the environment variable isn't necessary:

```python
# check to see if user requests a proxy connection.
# `get_proxy_context() returns the proxy context if the user requested use of one, else `None`.
proxy_context = get_proxy_context()
if proxy_context:
    # get the file as a bytes blob
    request = urllib.request.Request(url)
    with urllib.request.urlopen(request, context=proxy_context()) as fr:
    # ...
```

BobbyRBruce replied:
Or, `with urllib.request.urlopen(request, context=proxy_context) as fr:` rather.

BobbyRBruce commented on src/python/gem5/resources/downloader.py:
```diff
@@ -79,38 +80,20 @@ def _download(url: str, download_to: str, max_attempts: int = 6) -> None:
+                with urllib.request.urlopen(
+                    request, context=get_proxy_context()
+                ) as fr:
                     with tqdm.wrapattr(
                         open(download_to, "wb"),
                         "write",
```
I'm curious as to why we've got two different ways to download depending on whether a proxy was specified or not. The only difference appears to be that if proxy is specified `urlopen` is used while if it's not `urlretrieve` is used. If we used the `urlopen` approach for both, would that not work?

**This is not a blocker for this PR. Nor am I seriously asking for you to investigate. This isn't your code. I just think it's an odd design and I don't know the reason**.

powerjg replied:
This is my code, IIRC. I don't remember why. Hopefully I wrote something down in the commit message 😆

Harshil2107 committed:
stdlib: address requested changes

Change-Id: If78549c923518436006b0b302a5a073410b8a760
